en/dev_log/log_encryption.md

```diff
@@ -0,0 +1,179 @@
+
+::: warning
+Log encryption is not enabled by default in PX4 firmware builds.
+To use it you will need to build firmware with this feature enabled and then upload it to the flight controller (see instructions below).
+:::
+
+::: info
+The encryption algorithm used can be changed by modifying [SDLOG_ALGORITHM](../advanced_config/parameter_reference.md#SDLOG_ALGORITHM).
+However at time of writing, only XChaCha20 can actually be used (AES can be selected, but there is no implementation).
+:::
+
+## How ULog Encryption Works
+
+When the ULog file is first opened:
+
+- If [SDLOG_ALGORITHM](../advanced_config/parameter_reference.md#SDLOG_ALGORITHM) is set to default, a symmetric XChaCha20 key is generated and wrapped (encrypted) with RSA2048 public key encryption
+- This wrapped key is stored on the SD Card with the ".ulgk" suffix.
+
+While the ULog file is written to disk:
+
+- The generated (plaintext) key is used to encrypt the data blocks right before writing to disk.
+
+After the flight, there are two files:
+
+- `.ulogc` (ulog cipher): the actual log file data, encrypted
+- `.ulogk` (ulog wrapped key): the symmetric key, wrapped with rsa
+
+However, these files will both have the normal `.ulg` suffix if downloading from QGroundControl.
+
+## Creating a Flight Controller Build that contains Log Encryption
+
+Crypto uses large amounts of flash memory, and is therefore not included in PX4 builds by default (the exception is the make target `px4-fmu-v5_cryptotest`).
+Below we show how to add log encryption functionality to a flight controller target, using the `px4-fmu-v5` board as an example.
+
+Reference the `cryptotest.px4board` file in `boards/px4/fmu-v5` for the arguments you will need to add to the `.px4board` file associated with the flight controller you are targeting.
+You could also copy and paste `cryptotest.px4board` to the other flight controller's `boards/` location if you want this to be a separate build entirely (in that case, your `make` command would include `_cryptotest` at the end to add log encryption).
+
+### Crypto .px4board arguments
+
+| Argument                     | Description                                                                                                                        |
+| ---------------------------- | ---------------------------------------------------------------------------------------------------------------------------------- |
+| CONFIG_BOARD_CRYPTO          | Requirement for log encryption functionality to be enabled.<br />= `y`: Enable log encryption.<br />= `n`: Disable log encryption. |
+| CONFIG_DRIVERS_SW_CRYPTO     | Enables the built-in PX4 crypto backend library to handle logfile encryption.<br />= `y`: Enable<br />= `n`: Disable               |
+| CONFIG_DRIVERS_STUB_KEYSTORE | Enables the built-in PX4 stub keystore driver.<br />= `y`: Enable<br />= `n`: Disable                                              |
```

Comment:
So these all need to be set to y for the functionality to work? Aren't they all Y if you copy the values from `cryptotest.px4board`? I.e. why do you need to run kconfig then other than to set the location of the keys.

Reply:
Is it "Enable" or "include in firmware"? I assume the later - and once in firmware it is SDLOG_ALGORITHM that determines whether the feature is enabled or not? Also that this is default on if the module is installed?

Reply:
`So these all need to be set to y for the functionality to work? Aren't they all Y if you copy the values from cryptotest.px4board? I.e. why do you need to run kconfig then other than to set the location of the keys.`

The values set in kconfig add CONFIG_CRYPTO=y and CONFIG_CRYPTO_RANDOM_POOL=y  to defconfig when saving. I know at least CONFIG_CRYPTO being enabled is necessary.

When writing this, I believe I misunderstood the purpose of kconfig. If we can just set these in the .px4board without doing kconfig to add them to defconfig, then kconfig isn't necessary. I was just replicating what was done in the px4_fmu_v5 build, as it also has CONFIG_CRYPTO=y and CONFIG_CRYPTO_RANDOM_POOL=y set in defconfig.

Reply:
> Is it "Enable" or "include in firmware"? I assume the later - and once in firmware it is SDLOG_ALGORITHM that determines whether the feature is enabled or not? Also that this is default on if the module is installed?

I also think the latter.

Reply:
> When writing this, I believe I misunderstood the purpose of kconfig. If we can just set these in the .px4board without doing kconfig to add them to defconfig, then kconfig isn't necessary.

I _think_ you're right - Kconfig is a GUI tool for editing config files. I think the structure is fine, we just have to say that this is what we are doing/these are the alternatives

Reply:
I think the `CONFIG_DRIVERS_STUB_KEYSTORE` text "includes the built-in PX4 stub keystore driver" is correct - but it doesn't tell us much. I would guess that PX4 is suppoed to have a secure keystore, perhaps in a secure part. However that depends on hardware, so by default the keystore probably must maps to somewhere in the firmware or the SITL filesystem.